src/ucp/am/eager.inl

```diff
@@ -50,6 +50,10 @@ static void ucp_am_eager_zcopy_completion(uct_completion_t *self)
                                           send.state.uct_comp);
 
     ucs_assert(req->send.msg_proto.am.header.reg_desc != NULL);
+    if (ucs_unlikely(req->flags & UCP_REQUEST_FLAG_USER_HEADER_COPIED)) {
+        ucp_am_release_user_header(req);
```

Comment:
can we do it in ucp_am_eager_zcopy_pack_user_header?
so we release the mpool element as soon as we copy it to reg_desc
IMO it will be more self-documenting

Reply:
It might require to set `UCP_REQUEST_FLAG_USER_HEADER_COPIED` flag back to zero because protocol might change more than once if I'm not mistaken. 
For example:
[Some initial protocol] - Allocate ucx buffer in ptr. copy hdr to ucx buffer in ptr.
Zcopy - pack and release buffer in ptr. zcopy reset will not allocate new buffer in ptr because `UCP_REQUEST_FLAG_USER_HEADER_COPIED` flag is on
[New protocol selection]

If I'm not mistaken I don't like the idea of setting the flag back to zero.

Reply:
yes, i meant setting UCP_REQUEST_FLAG_USER_HEADER_COPIED to 0 after we call ucp_am_release_user_header (which is after we copied the header to reg_desc).

Reply:
We can do that, although I don't like the idea of setting the flag back to zero...
We can also take a different approach, we basically need to keep an invariant, header resides in only one buffer, reg_desc when proto is zcopy (after pack) or .ptr when it's not (after reset).

- pack func will copy the header to reg_desc and free .ptr if needed ( `UCP_REQUEST_FLAG_USER_HEADER_COPIED` is set).
- reset func will ignore `UCP_REQUEST_FLAG_USER_HEADER_COPIED` flag because of the invariant, it will allocate buffer if `UCP_AM_SEND_FLAG_COPY_HEADER` is set and copy the header from reg_desc to .ptr
- abort func will be the old `ucp_proto_request_zcopy_abort` (no need to release the header because of the invariant)

WDYT?

Reply:
As far as I understood @yosefe suggestion, it allows for a clearer and self-explanatory code, also indeed makes more sense to release immediately after copy in such case IMO.